---
title: "Colon T6B derepression CDF using merged peaks"
author: "Bing Shui"
date: "`r Sys.Date()`"
output: 
  html_document:
    toc: yes
    toc_float: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Prepare libraries
```{r message=FALSE, warning=FALSE}
library(DESeq2)
library(data.table)
library(gdata)
library(rtracklayer)
library(BSgenome)
library(VennDiagram)
library(ggplot2)
library(biomaRt)
library(RColorBrewer)
library(readr)
library(ensembldb)
library(EnsDb.Mmusculus.v79)
library(AnnotationDbi)
library(tidyverse)
library(plotly)
```
# Load dataset information
```{r}
dir.create("PDF_figure/T6B_derepression_CDF_merged", showWarnings = FALSE)

peaks.mir <- readRDS("Datafiles/merged-peaks-mirs-200-12232019-withID.rds")
mirs.peaks <- readRDS("Datafiles/miRNA-merged-peaks-list-12232019-withIDs.rds")

T6B_DGE <- read_csv("~/OneDrive - Harvard University/Haigis Lab/Projects/Halo-Ago2/Halo-Ago-KRas/Raw Data/RNA-Seq/T6B mouse colon/Result/Differential Analysis.csv")

colnames(T6B_DGE)[1] <- "gene"
colnames(T6B_DGE)[3] <- "log2FC"
```

# CDF plotting
## Set-up CDF plotting function
```{r}
plotCDF.ggplot <- function(gene.counts, gene.sets, gene.set.labels,
                           col = "", linetype = "", xlim = c( -1.0, 1.3 ),
                           legend.size = 22, axistitle.size = 22, title = "Fold change log2 (Dicer KO/WT)",
                           legend.pos = c(0.7, 0.18)) {
  require(ggplot2)
  df.log2fc <- gene.counts[,c("gene", "log2FC")]
  #rownames(df.log2fc) <- df.log2fc$gene
  if (length(gene.sets) != length(gene.set.labels)){
    return("Length of gene sets doesn't match labels")
  }
  target.expr <- df.log2fc[df.log2fc$gene %in% gene.sets[[1]],]
  for (i in 2:length(gene.sets)){
    target.expr <- rbind(target.expr, df.log2fc[df.log2fc$gene %in% gene.sets[[i]],])
  }

  gene.set.counts <- c()
  for (j in 1:length(gene.sets)){
    gene.set.counts <- c(gene.set.counts, sum(df.log2fc$gene %in% gene.sets[[j]]))
  }
  
  target.expr$Category <- rep(gene.set.labels, gene.set.counts)
  target.expr$Category <- factor(target.expr$Category, levels = gene.set.labels)
  
  log2FC.values <- lapply(gene.sets, function(gene.set) {
    gene.counts[gene.counts$gene %in% gene.set,]$log2FC
  })

  ks.pvals <- lapply(log2FC.values,
                     function(log2FCs) {
                       ks.test(log2FCs, log2FC.values[[1]])$p.value
                     })
  
  p <- ggplot( target.expr, aes( x = log2FC, colour = Category ) ) +
  stat_ecdf( geom = 'step', aes( colour = Category, linetype = Category ), lwd = 1 ) +
  scale_color_manual( values = col, labels = sprintf( "%s (%d)", gene.set.labels, gene.set.counts ) ) +
  scale_linetype_manual( values = linetype, labels = sprintf( "%s (%d)", gene.set.labels, gene.set.counts ) ) +
  # xlim() will remove data points; Be careful in the future
  coord_cartesian( xlim = xlim ) + xlab(title) + ylab('CDF') +
  theme_classic() + theme( legend.background = element_rect(fill = NA), 
                           legend.title = element_blank(), 
                           legend.position = legend.pos,
                           legend.text = element_text(size=legend.size),
                           legend.key.size = unit(1.5, 'lines'),
                           axis.title.x = element_text(size=axistitle.size, margin = margin(t = 10)),
                           axis.title.y = element_text(size=axistitle.size, margin = margin(r = 10)),
                           axis.text=element_text(size=20),
                           axis.line = element_line(size = 1),  #axis label size
                           axis.ticks.length = unit(0.3, "cm")) #increase tick length
  
  for (k in 2:length(gene.sets)){
    p <- p + annotate(geom = "text", x = -0.9, y = 1-0.08*(k-1), hjust = 0, 
                      label = sprintf("p = %.0e", ks.pvals[k]), 
                      colour = col[k], size = 8)
  }
  print(p)
}
```

## Plotting
```{r}
cols <- c(brewer.pal(name = "Set2", n = 8), brewer.pal(name = "Set3", n = 3))
plotCDF.ggplot(T6B_DGE,
               list(T6B_DGE$gene, peaks.mir$target_Ensembl_ID),
               c("All genes", "miRNAs over 200"),
               col = c("grey15", cols[1]),
               linetype = c(1, 1),
               title = "RNA_LFC(T6B/control)",
               legend.size = 15
               )

pdf("PDF_figure/T6B_derepression_CDF_merged/T6B_CDF_miRNAover200.pdf",
    height = 8,
    width = 10)
plotCDF.ggplot(T6B_DGE,
               list(T6B_DGE$gene, peaks.mir$target_Ensembl_ID),
               c("All genes", "miRNAs over 200"),
               col = c("grey15", cols[1]),
               linetype = c(1, 1),
               title = "RNA_LFC(T6B/control)",
               legend.size = 15
               )
dev.off()
```

```{r}
len <- sapply(mirs.peaks, function(x) length(x))
mirs.peaks <- mirs.peaks[order(-len)]
mirna <- names(mirs.peaks)

for (i in 1:10) {
  plotCDF.ggplot(T6B_DGE,
               list(T6B_DGE$gene, mirs.peaks[[i]]$target_Ensembl_ID),
               c("All genes", mirna[i]),
               col = c("grey15", cols[i]),
               linetype = c(1, 1),
               title = "RNA_LFC(T6B/control)",
               legend.size = 15
               )
}

pdf("PDF_figure/T6B_derepression_CDF_merged/T6B_CDF_top10miRNA.pdf",
    height = 8,
    width = 10)
for (i in 1:10) {
  plotCDF.ggplot(T6B_DGE,
               list(T6B_DGE$gene, mirs.peaks[[i]]$target_Ensembl_ID),
               c("All genes", mirna[i]),
               col = c("grey15", cols[i]),
               linetype = c(1, 1),
               title = "RNA_LFC(T6B/control)",
               legend.size = 15
               )
}
dev.off()
```

# CDF z-score 
## Prepare function
```{r}
zscore.cal <- function(genes = NA, index = NA, expression.dataset){
  if (sum(!is.na(genes)) > 0 && is.na(index)){
    gene.set <- expression.dataset[expression.dataset$gene %in% genes,]
  } else if (!is.na(index)) {
    gene.set <- expression.dataset[index,]
  } else {
    print("Please input gene names or row index.")
  }
  mu <- mean(expression.dataset$log2FC)
  Sm <- mean(gene.set$log2FC)
  m <- length(genes)
  SD <- sd(expression.dataset$log2FC)
  z <- (Sm-mu)*sqrt(m)/SD
  return(z)
}
```

## Calculation
```{r}
zscores.all <- as.data.frame(sapply(mirs.peaks,
                              function(targets){
                                  zscore.cal(genes = targets$target_Ensembl_ID, expression.dataset = T6B_DGE)
                            }))
colnames(zscores.all) <- c("z.score")
lens <- as.data.frame(sapply(mirs.peaks, function(x) length(x)))
colnames(lens) <- "N"
zscores.all <- cbind(zscores.all, lens$N)

mirna.family.DGE <- readRDS("Datafiles/mirna-counts-deseq-by-family-12232019.rds")
zscores.all <- as_tibble(zscores.all, rownames = "miR.family")
zscores.all <- inner_join(zscores.all, mirna.family.DGE[,c(1,7)], by = c("miR.family" = "miR.family"))
colnames(zscores.all)[3] <- "N"
```

## Plotting
```{r}
p1 <- ggplot(zscores.all, aes(x = z.score, y = log2(baseMean), size = N, label = miR.family)) +
  geom_point(colour = "#15CD40", alpha = 0.6) +
  xlab("T6B RNA-Seq Z-score") +
  ylab("miRNA family abundance") +
  theme_bw() +
      theme(panel.border = element_rect(),
      panel.background = element_blank(),
      panel.grid.major = element_line(), 
      panel.grid.minor = element_line(),
      axis.title.x = element_text(size=14, margin = margin(t = 10)),
      axis.title.y = element_text(size=14, margin = margin(r = 10)),
      axis.text = element_text(size=10),
      axis.line.y = element_blank(),
      axis.line.x = element_blank(),
      axis.ticks.x = element_line(size = 0.5),
      axis.ticks.y = element_line(size = 0.5),
      plot.title = element_text(hjust = 0.5, size = 14, face = "bold"))

cor.test(zscores.all$z.score, log2(zscores.all$baseMean))

pdf("PDF_figure/T6B_derepression_CDF_merged/T6B_CDF_zscore.pdf",
    height = 4,
    width = 6)
p1
dev.off()

ggplotly(p1)
```

# Session Info
```{r}
sessionInfo()
```
